docs/make.jl

```diff
@@ -18,43 +18,96 @@ using GAP.Markdown
 ## of the function `Documenter.Selectors.order` for these subtypes,
 ## and executing the step for the subtype `T` means to call
 ## `Documenter.Selectors.runner` with `T` and the document object.
-##
-## The code was adapted from the package
-## https://github.com/ali-ramadhan/DocumenterBibliographyTest.jl,
-## the hint to use this approach came up in the discussion of
-## https://github.com/JuliaDocs/Documenter.jl/issues/1343.
 
 ## Declare a new subtype for our step,
 ## the evaluation of external references.
 abstract type ExternalReference <: Documenter.Builder.DocumentPipeline end
 
-## Execute our step after the computation of cross-references (3.0)
-## and before the check of the document (4.0).
-Documenter.Selectors.order(::Type{ExternalReference}) = 3.1
+## Distinguish different versions of Documenter.jl.
+if isdefined(Documenter, :Document)
+  # version at least 1.0
+  # The code was adapted from `expand_citations`
+  # in DocumenterCitations.jl/src/citations.jl.
+  Document = Documenter.Document
 
-## When executing our step, print an info line
-## and call the function that does the work.
-function Documenter.Selectors.runner(::Type{ExternalReference}, doc::Documenter.Documents.Document)
-    @info "ExternalReference: building external references."
-    compute_external_references(doc)
-end
+  # Execute our step before the computation of cross-references (3.0)
+  # because otherwise the syntax of external references leads to errors.
+  Documenter.Selectors.order(::Type{ExternalReference}) = 2.99
+
+  function compute_external_references(doc::Document)
+    for (src, page) in doc.blueprint.pages
+      @debug "ExternalReference: compute for entries in $(src)"
+      empty!(page.globals.meta)
+      compute_external_reference(doc, page, page.mdast)
+    end
+  end
+
+  function compute_external_reference(doc::Document, page, mdast::Documenter.MarkdownAST.Node)
+    for node in Documenter.AbstractTrees.PreOrderDFS(mdast)
+      if node.element isa Documenter.DocsNode
+        # The docstring AST trees are not part of the tree of the page, so
+        # we need to expand them explicitly
+        for (docstr, meta) in zip(node.element.mdasts, node.element.metas)
+          compute_external_reference(doc, page, docstr)
+        end
+      elseif node.element isa Documenter.MarkdownAST.Link
+        compute_external_reference(node, page.globals.meta, page, doc)
+      end
+    end
+  end
 
-function compute_external_references(doc::Documenter.Documents.Document)
+  function compute_external_reference(node::Documenter.MarkdownAST.Node, meta, page, doc)
+    # Do something only if the current element is a `MarkdownAST.Link`.
+    link = node.element
+    isa(link, Documenter.MarkdownAST.Link) || return true
+    url = link.destination
+    startswith(url, "GAP_ref") || return true
+    @info "Computing external reference: $(url)."
```

Comment:
This function seems to be mostly identical to the old `compute_external_reference(link, meta, page, doc)` method below. Only instead of nested `if` blocks this one here aborts early via `|| return true` ; and of course `links.url` vs. `links.destination`.

Can't we extract the main body of this into a helper taking a single argument, `url`, and which returns either `nothing` or else the modified `url` value.

Then the two methods can both call this helper, and if it returns a value `!== nothing` they assign it to `link.destination` resp. `link.url`.

Reply:
O.k.
I think the point is that having such a helper function is anyhow a good idea, not just that it avoids a (temporary) code duplication.